### PR TITLE
Add design principles vignette

### PR DESCRIPTION
**Pull request by joshwlambert**
This PR addresses #28 by adding a design principles vignette. This vignette outlines the scope of the package, the function signatures and naming convention, as well as the function return types, it covers development design decisions and some explanation to package dependencies.

### CODE REVIEW

File: vignettes/design-principles.Rmd [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+## Output
+
+Functions with the name `probability_*()` return a single `numeric`. Functions with the name `proportion_*()` return a `<data.frame>` with as many rows as combinations of input values (see `expand.grid()`). The consistency of simple well-known data structure makes it easy for users to apply these functions in different scenarios.
+
+The distribution likelihoods return a vector of `numeric`s of equal length to the input vector. This is the same behaviour as the base R distribution functions.
+
+## Design decisions
+
+- `proportion_*()` functions return a `<data.frame>` with the proportion column(s) containing `character` strings, formatted with a percentage sign (`%`). It was reasoned that {superspreading} is most likely used either as a standalone package, or at the terminus of a epidemiological analysis pipeline, and thus the outputs of {superspreading} functions would not be passed into other functions. If this were the case then the proportions would be better as `numeric`.
```
Comment (sbfnk) [inline vignettes/design-principles.Rmd:37]: The rationale makes sense but the downside is that you can't plot them - perhaps make it an option?

**Review by sbfnk — COMMENTED** [review]
Thanks, @joshwlambert. This looks great.

Is it worth at this point reconsidering whether it would make merge this and `<epichains>` into a single package that revolves around analysis and interpretation of infectious disease data with branching process models? The main reason I'm thinking that this could make sense is that the two have somewhat circular dependencies: the mixture probability distributions implemented here would be useful in `<epichains>` (which already implements a gamma/Borel mixture) as would the functions for providing secondary analysis given parameters of a negative binomial - whereas this package itself depends on `<epichains>`/`<bpmodels>` for simulation.

If I remember correctly we previously decided to keep them separate as <bpmodels> was transitioning into <epichains> when this package came into existence but as this transition is mostly complete it might be a good time to come back to this.

cc @jamesmbaazam and @adamkucharski for thoughts.

**Comment by adamkucharski** [comment]
Agree makes sense to reconsider whether better as single package as a 'one stop shop' for branching analysis – currently `superspreading` is more focused on offspring distributions and analytical solutions vs branching simulation in `epichains`, but as you say @sbfnk there is potential for benefits of shared functionality across several aspects.

**Comment by sbfnk** [comment]
Great. Just to note that I've edited my initial comment having realised that none of the package names had been rendered - hopefully makes a bit more sense now.

**Comment by joshwlambert** [comment]
Thanks for the comments. I'm open to the merger. I think for now it would benefit both packages to reach stability and then be merged. Therefore, I propose we wait until after {epichains} has been reviewed and released and another {superspreading} release, and then we can coordinate a possible merge.

File: vignettes/design-principles.Rmd [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+This document is primarily intended to be read by those interested in understanding the code within the package and for potential package contributors.
+
+## Scope
+
+The {superspreading} package aims to provide a range of summary metrics to characterise individual-level variation in disease transmission. These include calculating the probability an outbreak becomes an epidemic (`probability_epidemic()`), or conversely goes extinct (`probability_extinct()`), the probability an outbreak can be contained (`probability_contain()`), the proportion of cases in cluster of a given size (`proportion_cluster_size()`), and the proportion of cases that cause a proportion of tranmsision (`proportion_transmission()`).
+
+The other aspect of the package is to provide likelihoods for distribution models to estimate heterogeneity in individual-level disease transmission that are not available in R (i.e. base R). At present we include two models: Poisson-lognormal (`dpoislnorm()` & `ppoislnorm()`) and Poisson-Weibull (`dpoisweibull()` & `ppoisweibull()`) distributions.
```
Comment (jamesmbaazam) [inline vignettes/design-principles.Rmd:25]: Are these probability distributions or likelihoods?

File: vignettes/design-principles.Rmd (hunk 1 above)
Comment (joshwlambert) [reply vignettes/design-principles.Rmd:37]: This has been logged as issue #69. I'll make the change in a dedicated PR.

File: vignettes/design-principles.Rmd [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+Functions with the name `probability_*()` return a single `numeric`. Functions with the name `proportion_*()` return a `<data.frame>` with as many rows as combinations of input values (see `expand.grid()`). The consistency of simple well-known data structure makes it easy for users to apply these functions in different scenarios.
+
+The distribution likelihoods return a vector of `numeric`s of equal length to the input vector. This is the same behaviour as the base R distribution functions.
+
+## Design decisions
+
+- `proportion_*()` functions return a `<data.frame>` with the proportion column(s) containing `character` strings, formatted with a percentage sign (`%`). It was reasoned that {superspreading} is most likely used either as a standalone package, or at the terminus of a epidemiological analysis pipeline, and thus the outputs of {superspreading} functions would not be passed into other functions. If this were the case then the proportions would be better as `numeric`.
+
+- The distribution likelihoods are vectorised (i.e. wrapped in `Vectorize()`). This enables them to be used identically to base R distribution functions.
```
Comment (jamesmbaazam) [inline vignettes/design-principles.Rmd:39]: The use of "distribution likelihoods" is potentially confusing. Could you clarify which one it is?

File: vignettes/design-principles.Rmd (hunk 2 above)
Comment (jamesmbaazam) [inline vignettes/design-principles.Rmd:25]: Does the package implement likelihoods or probability distributions?

File: vignettes/design-principles.Rmd [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+The aim is to restrict the number of dependencies to a minimal required set for ease of maintenance. The current hard dependencies are:
+
+* {stats}
+* [{checkmate}](https://CRAN.R-project.org/package=checkmate)
+* [{bpmodels}](https://github.com/epiverse-trace/bpmodels)
+
+{stats} is distributed with the R language so is viewed as a lightweight dependency, that should already be installed on a user's machine if they have R. {checkmate} is an input checking package widely used across Epiverse-TRACE packages.  {bpmodels} is used to simulate a single-type branching process model to calculate the probability of containment (`probability_contain()`), as mentioned above, branching process simulations are beyond the scope of {superspreading}.
+
+Suggested dependencies (not including package documentation ({knitr}, {bookdown}, {rmarkdown}), testing ({spelling} and {testthat}), and plotting ({ggplot2})) are: [{epiparameter}](https://github.com/epiverse-trace/epiparameter), used to easily access epidemiological parameters from the package's library, and [{fitdistrplus}](https://CRAN.R-project.org/package=fitdistrplus), used for model fitting methods.
```
Comment (jamesmbaazam) [inline vignettes/design-principles.Rmd:55]: I like this section.

File: vignettes/design-principles.Rmd (hunk 2 above)
Comment (joshwlambert) [reply vignettes/design-principles.Rmd:25]: Thanks for raising. They are PDFs and CDFs. I can reword it to make this clear, they are used to compute the likelihood given:

$\[ L(\theta; x_1, x_2, \ldots, x_n) = f(x_1; \theta) \times f(x_2; \theta) \times \ldots \times f(x_n; \theta) \]$.

File: vignettes/design-principles.Rmd [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+---
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+```
+
+This vignette outlines the design decisions that have been taken during the development of the {superspreading} R package, and provides some of the reasoning, and possible pros and cons of each decision.
```
Comment (jamesmbaazam) [inline vignettes/design-principles.Rmd:17]: I don't think I've noticed the pros and cons mentioned here. One potentially consideration is #69.

Comment (joshwlambert) [reply vignettes/design-principles.Rmd:17]: Thanks, will mention in issue #69 and update the design vignette in that PR.

**Review by jamesmbaazam — APPROVED** [review]
Thanks, Josh. I like that the document is very clear about the scope of the package. I've always wondered if this document is also a good place to compare other related packages.

File: vignettes/design-principles.Rmd (hunk 2 above)
Comment (jamesmbaazam) [reply vignettes/design-principles.Rmd:25]: I agree that rewording would clarify this.

**Comment by jamesmbaazam** [comment]
> Thanks for the comments. I'm open to the merger. I think for now it would benefit both packages to reach stability and then be merged. Therefore, I propose we wait until after {epichains} has been reviewed and released and another {superspreading} release, and then we can coordinate a possible merge.

Thanks for your inputs, all. I do think that merging the two packages would be a good idea because they have enough overlap. Additionally, a single package would be easier to maintain since any breaking changes in {epichains} would affect {superspreading}. If the agreement is to merge it, I would suggest holding off another release of {superspreading} and rather merging it into {epichains} before the v0.1.0 release in the new year. We can then jointly supersede {bpmodels} and {superspreading}.

File: vignettes/design-principles.Rmd (hunk 3 above)
Comment (joshwlambert) [reply vignettes/design-principles.Rmd]: Resolved in f2aef3db14bf4ffeefbb474f2ad0bf5e27d3c330.

**Comment by joshwlambert** [comment]
Thanks for all the input. I will merge this PR once all checks pass. I think the discusson on merging should be continued elsewhere.
